[GH-317] Correct size for Linux tray icons

elsiehupp opened the pull request:
## Summary
The existing Mattermost tray icons are sized inconsistently with other Linux tray icons. This PR adds SVG versions of the Linux tray icons (which apparently didn’t exist previously) and PNG exports with new dimensions based on the [GNOME Symbolic Icon Design Guidelines](https://developer.gnome.org/hig/stable/icon-design.html.en); that is, a 16px x 16px grid with 1px margins. No further changes to the icon design were made (though the appearance might be slightly different due to having to recreate the SVG masters).

## Ticket Link
Fixes https://github.com/mattermost/desktop/issues/317.

## Checklist
- [x] Has UI changes
- [x] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
- [x] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-cobtributor-agreement/)

#### Device Information
> I tried testing this on vanilla GNOME 3.38, and the AppIndicator doesn’t work, period, so I don’t know what it looks like.

#### Screenshots
> See above.

#### Release Note
```release-note
Fixed tray icon size on Linux.
```

mattermod commented:
Hello @elsiehupp,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

elsiehupp commented:
It’s a complete mystery to me why the CI is failing, considering I didn’t change any existing files other than PNG images.

svelle commented:
![Screenshot from 2021-05-07 10-26-12](https://user-images.githubusercontent.com/7526550/117421404-cbeb3080-af1e-11eb-9081-4708b95c5241.png)

Builds without issues on Fedora 34 and displays tray icon as supposed (Gnome 40 + Appindicator Extension)

Willyfrog commented:
> ![Screenshot from 2021-05-07 10-26-12](https://user-images.githubusercontent.com/7526550/117421404-cbeb3080-af1e-11eb-9081-4708b95c5241.png)
> 
> Builds without issues on Fedora 34 and displays tray icon as supposed (Gnome 40 + Appindicator Extension)

Ah! I was going to post a screenshot from ubuntu, but looks good as well

matthewbirtch commented:
@Willyfrog how can I test this to do a visual review?

Willyfrog commented:
@matthewbirtch you'll need a linux machine and grab one of the builds from circleci, for this pr those are located at https://app.circleci.com/pipelines/github/mattermost/desktop/2216/workflows/1c66f799-1741-4933-b1bd-c24ee3442ee5/jobs/9893/artifacts

elsiehupp commented:
FYI if you’d like me to make any tweaks to the placement of the indicator dot, that would be fairly straightforward for me to do.

For example, in this screenshot IMHO maybe the dot could stand to be a little bit further down and to the right, idk:

![Screenshot from 2021-05-07 11-37-14](https://user-images.githubusercontent.com/1515906/117430914-e62a0c00-af28-11eb-9fdc-befd2de660f8.png)

elsiehupp commented:
I went ahead and made the same changes to the scaling and margin on [a copy of the Figma](https://www.figma.com/file/b8suZ70gAnTn0abdF3Undz/MM-30584-Desktop-App-Icon-Update-Copy?node-id=9%3A1170). I’m not sure how to do a pull request there, though.

One advantage with working directly in SVG is that it can integrate with Git because it’s literally just a text document. (To be fair, I actually work in [Affinity](https://affinity.serif.com/designer/), but I know how to do fairly clean SVG exports.)

In addition to making the changes on Figma, I downloaded an SVG export and superimposed it on my version. The dot was more or less identical (though it was on a slightly different pixel grid because I added it after rather than before scaling the icon from 16x16 to 14x14), and the main difference is that the cutout is much larger. (Also the “dark” icons on Figma use a lighter gray rather than translucent black, which is how the previous PNG exports had done it.)

One thought I had was that if the main icon is 14x14, but the canvas is 16x16, the dot could be 1px further down and to the right, which among other things would allow a larger cutout without cutting out of the dial hand. How does this look?

![MenuIconMention@2x-dark](https://user-images.githubusercontent.com/9206310/117851562-6ff11680-b254-11eb-8727-f61c0eca68b0.png) ![MenuIconMention@2x-light](https://user-images.githubusercontent.com/9206310/117851477-5d76dd00-b254-11eb-9e8f-d432f4e0e941.png) ![MenuIconUnread@2x-dark](https://user-images.githubusercontent.com/9206310/117852328-366cdb00-b255-11eb-8436-8b59e22615c0.png) ![MenuIconUnread@2x-light](https://user-images.githubusercontent.com/9206310/117852550-6c11c400-b255-11eb-939e-05908ec1a6ac.png)



It would be important to test this in a build, though, since Electron does weird things, and I’d want to make sure putting the dot in 1px margin is in fact safe to do.

elsiehupp commented:
Oh, and if you prefer the version with dot further down and to the right I can copy the design over to Figma, as well.

matthewbirtch commented:
> One thought I had was that if the main icon is 14x14, but the canvas is 16x16, the dot could be 1px further down and to the right, which among other things would allow a larger cutout without cutting out of the dial hand. How does this look?

Thanks @elsiehupp. Let's leave it as designed for now. I'd prefer not to scale down the main icon. Plus, what has been designed in Figma is already implemented in MacOS so it's best to keep things consistent.

Given this, do you need anything else to finish this one off?

elsiehupp commented:
> Given this, do you need anything else to finish this one off?

I just replaced the PNGs with fresh exports from my Figma fork, and I used the default filenames derived from the Figma labels. Because I changed the filenames, there is some chance that the related code (which I also changed) doesn’t work.

I also added a Markdown document pointing to the Figma, because the canonical source of the icons wasn’t clear to me when I did me initial work. This sort of readme could probably be done more elegantly (with proper deep-links), and it could probably stand to be replicated elsewhere in the GitHub.

Once everything is working here you could copy the changes from [my Figma fork](https://www.figma.com/file/b8suZ70gAnTn0abdF3Undz/MM-30584-Desktop-App-Icon-Update-Copy?node-id=9%3A1170) over to the Figma master. I removed all but the 16px sizes of the Linux tray icons because that’s the only size Electron supports, and the `@2x` versions of the exports are freebies with Figma, so they don’t need to be manually created.

elsiehupp commented:
> Let's leave it as designed for now. I'd prefer not to scale down the main icon. Plus, what has been designed in Figma is already implemented in MacOS so it's best to keep things consistent.

What I did with the updated version in the new squashed commit was I scaled the existing 16x16 Linux tray icons on Figma down to 14x14 but left the export slices at 16x16. This is because that’s just what Electron expects on Linux, and as far as I know is necessary for fixing Issue https://github.com/mattermost/desktop/issues/317.

Regarding slightly offsetting the dot, that could easily be done in a subsequent pull request (with corresponding changes on Windows and macOS, as well). You can still test the PNGs attached to my previous comment if you feel so inclined.

elsiehupp commented:
If you want to support a wider variety of tray icon sizes on Linux, and/or you want to be able to use SVGs for tray icons, you can reopen my [Chromium bug report](https://bugs.chromium.org/p/chromium/issues/detail?id=1167601) and ask the Chromium developers to fix the underlying problem.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

elsiehupp commented:
What exactly are we waiting for here? Let me know if there’s anything you need me to do. Thanks!

Willyfrog commented:
@matthewbirtch is there anything you need to do the ux review?

matthewbirtch commented:
@elsiehupp I got a little lost and confused in your last few comments. Can you provide screenshots of the latest? I'm unable to get this installed on my Linux VM right now.

elsiehupp commented:
> @elsiehupp I got a little lost and confused in your last few comments. Can you provide screenshots of the latest? I'm unable to get this installed on my Linux VM right now.

I’m not running Ubuntu at the moment (thought I’m planning on reinstalling it in a multi-boot environment), so I haven’t been able to test the build, per se. The most recent version of the PNGs (which would be used in the Linux build) are from the Figma masters but scaled to the size Electron expects for AppIndicators. Anyone here with a working copy of Ubuntu should be able to test the build. (The build works on other distress, but for whatever reason the AppIndicator currently only works on Ubuntu.)

In addition to replacing the PNGs, I changed the code so that the PNGs use the filenames Figma gives them (presumably based on layer names or something), and I added a Markdown README file to point future contributors to the Figma source, since it isn’t obvious otherwise. (I don’t know how to deep-link within Figma, but it could be useful to have deep links to specific places in the Figma wherever there are exports in the GitHub repo, since Figma doesn’t integrate nicely with Git.)

Finally, and this isn’t necessary for the build itself, I modified a copy of the Figma so that it has only the correct size (14x14 and 14x14@2x) for the Linux AppIndicator export. I don’t know how to do a pull request on Figma, but basically you or someone else would integrate my changes into the master branch there. You can find my Figma fork [here](https://www.figma.com/file/b8suZ70gAnTn0abdF3Undz/MM-30584-Desktop-App-Icon-Update-Copy?node-id=9%3A1170).

As an aside, someone here could reopen [my Chromium bug report](https://bugs.chromium.org/p/chromium/issues/detail?id=1167601) if you’d like to be able to use proper SVG AppIndicators on Linux. Using SVG allows the operating system to tint the file to the precise shade determined by the OS theme. (In other words, rather than having separate black and white exports, it would change the color of the main geometry to off-black or off-white while leaving the dot the correct accent color.)

elsiehupp commented on src/assets/linux/ICONS.md:
```diff
@@ -0,0 +1,3 @@
+## Figma
+
+Canonical vector source images for icons can be found on [Figma](https://www.figma.com/file/7V9FN9l9c4zHIls2ZUCoPU/MM-30584-Desktop-App-Icon-Update?node-id=9%3A1170). The files on GitHub are exports.
```
This is the README pointing to the master on Figma. It would be better with a deep link, but I don’t know how to do that.

elsiehupp commented on src/main/tray/tray.js:
```diff
@@ -45,16 +45,16 @@ export function refreshTrayImages(trayIconTheme) {
-                normal: nativeImage.createFromPath(path.resolve(assetsDir, 'linux', theme, 'MenuIcon.png')),
-                unread: nativeImage.createFromPath(path.resolve(assetsDir, 'linux', theme, 'MenuIconUnread.png')),
-                mention: nativeImage.createFromPath(path.resolve(assetsDir, 'linux', theme, 'MenuIconMention.png')),
+                normal: nativeImage.createFromPath(path.resolve(assetsDir, 'linux', 'top_bar_dark_16.png')),
```
The changes in this file reflect the renamed PNG files (which I also moved from the subfolders to the parent folder because they have distinct names, now).

elsiehupp commented:
Oh, I figured out how to do the deep link:

1. I created a new “frame” around the objects (using the second item on the toolbar),
2. then selected the menu item Options > Frame Options > Resize to Fit,
3. renamed the frame “Linux Exports”, and,
4. with the frame selected,
   a. clicked “Share”,
   b. ticked the box labeled “Link to selected frame”, and
   c. clicked “Copy link”, giving me [this](https://www.figma.com/file/b8suZ70gAnTn0abdF3Undz/MM-30584-Desktop-App-Icon-Update-Copy?node-id=316%3A0).

Of course, I can’t use this link for the README because it links to my copy, not the master copy. If you follow the above steps (after merging my changes in Figma), you can create the link yourself. (The fact that Figma is separate from GitHub really complicates things here.)

matthewbirtch commented:
> I’m not running Ubuntu at the moment (thought I’m planning on reinstalling it in a multi-boot environment), so I haven’t been able to test the build, per se. The most recent version of the PNGs (which would be used in the Linux build) are from the Figma masters but scaled to the size Electron expects for AppIndicators. Anyone here with a working copy of Ubuntu should be able to test the build. (The build works on other distress, but for whatever reason the AppIndicator currently only works on Ubuntu.)

How were you able to test what it would look like without Ubuntu?

As mentioned, I'm not super familiar with Linux and I'm not able to get the builds to install properly in my Ubuntu VM.

elsiehupp commented:
> How were you able to test what it would look like without Ubuntu?

The PNGs are the same pixel dimensions as the previous set I had in this pull request—which the other reviewers were able to test—and they’re the dimensions known to be preferred by Electron for AppIndicators. The only thing that should be different is the precise size and outline of the colored dot.

Willyfrog commented:
@matthewbirtch we can talk in community and I can help you get it installed in your VM or try to figure out if we are missing something

elsiehupp commented on src/main/tray/tray.js:
```diff
@@ -45,16 +45,16 @@ export function refreshTrayImages(trayIconTheme) {
         const theme = trayIconTheme;
```
Let’s see if this does it...

matthewbirtch commented:
> @matthewbirtch we can talk in community and I can help you get it installed in your VM or try to figure out if we are missing something

After a software update to my VM, I was able to get the app build installed and all looks good from my standpoint. Sorry for all the back and forth @elsiehupp!

elsiehupp commented:
@matthewbirtch no worries! Were you able to figure out merging the changes over on Figma, as well?

matthewbirtch commented:
> @matthewbirtch no worries! Were you able to figure out merging the changes over on Figma, as well?

There is no way to merge in Figma like with github, but I've updated our icon files to reflect the scaling for Linux

Willyfrog reviewed (approved):
LGTM! thanks for working on this :)

Willyfrog commented:
ci errors doesn't seem related to the PR, merging
